src/codegen/capacity/estimators.ts

```diff
@@ -0,0 +1,173 @@
+import {JsExpression} from '@jsonjoy.com/util/lib/codegen/util/JsExpression';
+import {MaxEncodingOverhead, maxEncodingCapacity} from '@jsonjoy.com/util/lib/json-size';
+import type {CapacityEstimatorCodegenContext} from './CapacityEstimatorCodegenContext';
+import type {Type} from '../../type';
+
+type EstimatorFunction = (ctx: CapacityEstimatorCodegenContext, value: JsExpression, type: Type) => void;
+
+function normalizeAccessor(key: string): string {
+  // Simple property access for valid identifiers, bracket notation otherwise
+  if (/^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key)) {
+    return `.${key}`;
+  }
+  return `[${JSON.stringify(key)}]`;
+}
+
+export function estimateAnyCapacity(ctx: CapacityEstimatorCodegenContext, value: JsExpression, type: Type): void {
+  const codegen = ctx.codegen;
+  codegen.link('Value');
+  const r = codegen.var(value.use());
+  codegen.if(
+    `${r} instanceof Value`,
+    () => {
+      codegen.if(
+        `${r}.type`,
+        () => {
+          ctx.codegen.js(`size += ${r}.type.capacityEstimator()(${r}.data);`);
+        },
+        () => {
+          ctx.codegen.js(`size += maxEncodingCapacity(${r}.data);`);
+        },
+      );
+    },
+    () => {
+      ctx.codegen.js(`size += maxEncodingCapacity(${r});`);
+    },
+  );
+}
+
+export function estimateBooleanCapacity(ctx: CapacityEstimatorCodegenContext, value: JsExpression): void {
```

Comment:
For functions exportable from this module use short names, such as "bool" instead of "estimateBooleanCapacity":

```suggestion
export const bool = (ctx: CapacityEstimatorCodegenContext, value: JsExpression): void => {
```